beacon-chain/rpc/eth/beacon/handlers_validator.go

```diff
@@ -0,0 +1,375 @@
+		return
+	}
+	st, err := bs.Stater.State(ctx, []byte(stateId))
+	if err != nil {
+		shared.WriteStateFetchError(w, err)
+		return
+	}
+
+	isOptimistic, err := helpers.IsOptimistic(ctx, []byte(stateId), bs.OptimisticModeFetcher, bs.Stater, bs.ChainInfoFetcher, bs.BeaconDB)
+	if err != nil {
+		http2.HandleError(w, "Could not check optimistic status: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	blockRoot, err := st.LatestBlockHeader().HashTreeRoot()
+	if err != nil {
+		http2.HandleError(w, "Could not calculate root of latest block header: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	isFinalized := bs.FinalizationFetcher.IsFinalized(ctx, blockRoot)
+
+	rawIds := r.URL.Query()["id"]
+	ids, ok := decodeIds(w, st, rawIds, true /* ignore unknown */)
+	if !ok {
+		return
+	}
+	// return no data if all IDs are ignored
+	if len(rawIds) > 0 && len(ids) == 0 {
+		resp := &GetValidatorBalancesResponse{
+			Data:                []*ValidatorBalance{},
+			ExecutionOptimistic: isOptimistic,
+			Finalized:           isFinalized,
+		}
+		http2.WriteJson(w, resp)
+		return
+	}
+
+	bals := st.Balances()
+	var valBalances []*ValidatorBalance
+	if len(ids) == 0 {
+		valBalances = make([]*ValidatorBalance, len(bals))
+		for i, b := range bals {
+			valBalances[i] = &ValidatorBalance{
+				Index:   strconv.FormatUint(uint64(i), 10),
+				Balance: strconv.FormatUint(b, 10),
+			}
+		}
+	} else {
+		valBalances = make([]*ValidatorBalance, len(ids))
+		for i, id := range ids {
+			valBalances[i] = &ValidatorBalance{
+				Index:   strconv.FormatUint(uint64(id), 10),
+				Balance: strconv.FormatUint(bals[id], 10),
+			}
+		}
+	}
+
+	resp := &GetValidatorBalancesResponse{
+		Data:                valBalances,
+		ExecutionOptimistic: isOptimistic,
+		Finalized:           isFinalized,
+	}
+	http2.WriteJson(w, resp)
+}
+
+// decodeIds takes in a list of validator ID strings (as either a pubkey or a validator index)
+// and returns the corresponding validator indices. It can be configured to ignore well-formed but unknown indices.
+func decodeIds(w http.ResponseWriter, st state.BeaconState, rawIds []string, ignoreUnknown bool) ([]primitives.ValidatorIndex, bool) {
+	ids := make([]primitives.ValidatorIndex, 0, len(rawIds))
+	numVals := uint64(st.NumValidators())
+	for _, rawId := range rawIds {
+		pubkey, err := hexutil.Decode(rawId)
+		if err == nil {
+			if len(pubkey) != fieldparams.BLSPubkeyLength {
+				http2.HandleError(w, fmt.Sprintf("Pubkey length is %d instead of %d", len(pubkey), fieldparams.BLSPubkeyLength), http.StatusBadRequest)
+				return nil, false
+			}
+			valIndex, ok := st.ValidatorIndexByPubkey(bytesutil.ToBytes48(pubkey))
+			if !ok {
+				if ignoreUnknown {
+					continue
+				}
+				http2.HandleError(w, fmt.Sprintf("Unknown pubkey %s", pubkey), http.StatusBadRequest)
+				return nil, false
+			}
+			ids = append(ids, valIndex)
+			continue
+		}
+
+		index, err := strconv.ParseUint(rawId, 10, 64)
```

Comment:
if we aren't doing anything with errors for pubkey then maybe this should be first so it doesn't go through all the pubkey logic if failed